source/server/filter_chain_manager_impl.h

```diff
@@ -21,6 +28,71 @@ class FilterChainFactoryBuilder {
   buildFilterChain(const ::envoy::api::v2::listener::FilterChain& filter_chain) const PURE;
 };
 
+// FilterChainFactoryContextImpl is supposed to used by network filter chain.
+// Its lifetime must cover the created network filter chain.
+// Its lifetime should be covered by the owned listeners so as to support replacing the active
+// filter chains in the listener.
+class FilterChainFactoryContextImpl : public Configuration::FilterChainFactoryContext,
+                                      public Network::DrainDecision {
+public:
+  FilterChainFactoryContextImpl(Configuration::FactoryContext& parent_context,
+                                uint64_t filter_chain_tag);
+
+  // FilterChainFactoryContext
+  uint64_t filterChainTag() const override;
+
+  // DrainDecision
+  bool drainClose() const override;
+
+  // Configuration::FactoryContext
+  AccessLog::AccessLogManager& accessLogManager() override;
+  Upstream::ClusterManager& clusterManager() override;
+  Event::Dispatcher& dispatcher() override;
+  Network::DrainDecision& drainDecision() override;
+
+  Grpc::Context& grpcContext() override;
+  bool healthCheckFailed() override;
+  Tracing::HttpTracer& httpTracer() override;
+  Http::Context& httpContext() override;
+  Init::Manager& initManager() override;
+  const LocalInfo::LocalInfo& localInfo() const override;
+  Envoy::Runtime::RandomGenerator& random() override;
+  Envoy::Runtime::Loader& runtime() override;
+  Stats::Scope& scope() override;
+  Singleton::Manager& singletonManager() override;
+  OverloadManager& overloadManager() override;
+  ThreadLocal::SlotAllocator& threadLocal() override;
+  Admin& admin() override;
+  const envoy::api::v2::core::Metadata& listenerMetadata() const override;
+  envoy::api::v2::core::TrafficDirection direction() const override;
+  TimeSource& timeSource() override;
+  ProtobufMessage::ValidationVisitor& messageValidationVisitor() override;
+  Api::Api& api() override;
+  ServerLifecycleNotifier& lifecycleNotifier() override;
+  OptProcessContextRef processContext() override;
+  Configuration::ServerFactoryContext& getServerFactoryContext() const override;
+
+  Stats::Scope& listenerScope() override;
+
+private:
+  Configuration::FactoryContext& parent_context_;
+  uint64_t filter_chain_tag_;
+};
+
+class FilterChainManagerImpl;
+
+class ThreadLocalFilterChainManagerHelper : public ThreadLocal::ThreadLocalObject {
+public:
+  // The FCM which can be snapped by worker.
+  std::shared_ptr<FilterChainManagerImpl> filter_chain_manager_;
+  // The per worker listener which owns the thread local filter chain manager
+  Network::ListenerCallbacks* listener_;
+
+  // Below could be mutated by main thread. Worker thread should access with caution.
+  std::unique_ptr<Init::ManagerImpl> fcm_init_manager_;
+  std::unique_ptr<Init::WatcherImpl> fcm_init_watcher_;
+  std::unique_ptr<std::unordered_set<uint64_t>> filter_chains_trait_;
+};
```

Comment:
Can we please remove any FCM related things from this PR? AIUI the only thing being done in this PR is plumbing through the new filter chain factory context.

Reply:
Will do